demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
+        Each state supports common methods like get_status_message(), get_status_color(), allow_configuration_changes()
+        """
+        return self._state
+
+    def build_form():
+        """ Convert self.configuration_schema into a Django form """
+        pass
+
+
+class VmProvider(Resource):
+    """ A provider that can provision virtual machines """
+    class Meta:
+        proxy = True
```

Comment:
My idea here was that subclasses would store all their fields in the `configuration` JSONField inherited from the parent class. The advantage of that is that it allows us to define many different Resource subclasses, and change them as needed, without having to deal with database migrations every time we add a new Resource. It also allows the parent class to "see" the configuration of children and know when it's changed, because we only need to know when that one JSONField changes. If we use multi-table inheritance, then detecting changes to child configuration/fields becomes a little bit harder to maintain (need to use save signals or override the save method consistently, etc. and even then you can't easily tell if any values were actually changed or not).

Also, since we're on PostgreSQL, a `JSONField` can be very efficient and even allow database-level filtering on the field data.